### PR TITLE
consensus, eth: implement gossip network

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

File: consensus/istanbul/backend.go [hunk 1]
```diff
@@ -34,9 +34,12 @@ type Backend interface {
 	// EventMux returns the event mux in backend
 	EventMux() *event.TypeMux
 
-	// Broadcast sends a message to all validators
+	// Broadcast sends a message to all validators (include self)
```
Comment (yutelin) [inline consensus/istanbul/backend.go]: keep the comment?

File: consensus/istanbul/backend/engine.go [hunk 2]
```diff
@@ -481,25 +479,12 @@ func (sb *backend) APIs(chain consensus.ChainReader) []rpc.API {
...
 
 // HandleMsg implements consensus.Istanbul.HandleMsg
-func (sb *backend) HandleMsg(pubKey *ecdsa.PublicKey, data []byte) error {
+func (sb *backend) HandleMsg(addr common.Address, data []byte) error {
 	sb.coreMu.Lock()
 	defer sb.coreMu.Unlock()
 	if !sb.coreStarted {
 		return istanbul.ErrStoppedEngine
 	}
-	addr := crypto.PubkeyToAddress(*pubKey)
```
Comment (yutelin) [inline consensus/istanbul/backend/engine.go]: where do we verify the sender is in validator set now?

**Comment by yutelin** [comment]
LGTM. Need to run stress testing and scalability testing.

File: consensus/istanbul/backend/engine.go (hunk 2 above)
Comment (markya0616) [inline consensus/istanbul/backend/engine.go]: In our core engine.

File: consensus/istanbul/backend.go (hunk 1 above)
Comment (markya0616) [inline consensus/istanbul/backend.go]: OK

File: consensus/istanbul/core/roundchange.go [hunk 3]
```diff
@@ -67,7 +67,7 @@ func (c *core) sendRoundChange(round *big.Int) {
 }
 
 func (c *core) handleRoundChange(msg *message, src istanbul.Validator) error {
-	logger := c.logger.New("state", c.state)
+	logger := c.logger.New("state", c.state, "from", src.Address())
```
Comment (bailantaotao) [inline consensus/istanbul/core/roundchange.go]: src.Address().Hex()?

Comment (markya0616) [inline consensus/istanbul/core/roundchange.go]: OK

File: consensus/istanbul/core/handler.go [hunk 4]
```diff
@@ -84,12 +84,21 @@ func (c *core) handleEvents() {
...
 		case istanbul.FinalCommittedEvent:
 			c.handleFinalCommitted(ev.Proposal, ev.Proposer)
 		case backlogEvent:
 			// No need to check signature for internal messages
-			c.handleCheckedMsg(ev.msg, ev.src)
+			if err := c.handleCheckedMsg(ev.msg, ev.src); err == nil {
+				p, err := ev.msg.Payload()
+				if err != nil {
+					c.logger.Warn("Get message payload failed", "err", err)
+					return
```
Comment (tailingchen) [inline consensus/istanbul/core/handler.go]: use `continue` here?

Comment (markya0616) [inline consensus/istanbul/core/handler.go]: OK

File: eth/istanbul_handler.go [hunk 5]
```diff
@@ -39,13 +40,18 @@ const (
...
 )
 
 type istanbulProtocolManager struct {
 	*protocolManager
 
-	engine   consensus.Istanbul
-	eventSub *event.TypeMuxSubscription
+	engine         consensus.Istanbul
+	eventSub       *event.TypeMuxSubscription
+	recentMessages *lru.ARCCache // the cache of peers' messages
```
Comment (bailantaotao) [inline eth/istanbul_handler.go]: s/peers'/peer's/

Comment (markya0616) [inline eth/istanbul_handler.go]: ok

File: eth/istanbul_handler.go [hunk 6]
```diff
@@ -120,7 +130,28 @@ func (pm *istanbulProtocolManager) handleMsg(p *peer, msg p2p.Msg) error {
...
+		// Mark peer's message
+		ms, ok := pm.recentMessages.Get(addr)
+		var m *lru.ARCCache
+		if ok {
+			m, _ = ms.(*lru.ARCCache)
+		} else {
+			m, _ = lru.NewARC(inmemoryMessages)
+		}
+		m.Add(hash, true)
+		pm.recentMessages.Add(addr, m)
```
Comment (tailingchen) [inline eth/istanbul_handler.go]: Move it after L143. No need to add cache again if it already exists.

**Pull request merged**